activerecord/test/cases/connection_adapters/connection_handler_test.rb

```diff
@@ -66,6 +66,35 @@ def test_establish_connection_using_3_levels_config
         ENV["RAILS_ENV"] = previous_env
       end
 
+      def test_using_connection
+        klass2 = Class.new(Base) { def self.name; "klass2"; end }
+
+        assert_equal ActiveRecord::Base.retrieve_connection("ARUnit2Model"), klass2.using_connection("ARUnit2Model", &:connection)
+        assert_equal ActiveRecord::Base.connection, klass2.connection
+      end
+
+      def test_using_connection_thread_safety
+        klass2 = Class.new(Base) { def self.name; "klass2"; end }
+        outer_connection = nil
+
+        barrier_1 = Concurrent::CyclicBarrier.new(2)
+        barrier_2 = Concurrent::CyclicBarrier.new(2)
+        [
+          Thread.new do
+            klass2.using_connection("ARUnit2Model") do |klass|
+              barrier_1.wait
+              barrier_2.wait(0.1)
```

Comment:
Reposting because GitHub erroneously collapsed the original as outdated:

@matthewd You used Concurrent::CyclicBarrier in dae525e, so I was hoping you could help. This line really isn't much better than sleep 0.1. What I'm trying to do is give outer_connection = klass2.connection a chance to start running while still in this block, but that call will actually wait until exiting this block because of the moitor I use, which is exactly what this test is ensuring. This means that removing the timeout results in deadlock. Any thoughts on how to improve this?